### ISSUE TITLE
How to use odoo enterprise addons with doodba

### ISSUE THREAD

**Issue by AungKoKoLin1997** [Tecnativa/doodba-copier-template#495]
I want to add enterprise addons for my testing environment and I am getting error message.
This is my `repos.yaml`.

```
# See https://github.com/Tecnativa/doodba#optodoocustomsrcreposyaml
./odoo:
  defaults:
    # Shallow repositories ($DEPTH_DEFAULT=1) are faster & thinner
    # You may need a bigger depth when merging PRs (use $DEPTH_MERGE
    # for a sane value of 100 commits)
    depth: $DEPTH_DEFAULT
  remotes:
    ocb: https://github.com/OCA/OCB.git
    odoo: https://github.com/odoo/odoo.git
    openupgrade: https://github.com/OCA/OpenUpgrade.git
  target: ocb $ODOO_VERSION
  merges:
    - ocb $ODOO_VERSION
    # Example of a merge of the PR with the number <PR>
    # - oca refs/pull/<PR>/head
./private:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    enterprise: https://github.com/odoo/enterprise.git
  target: enterprise $ODOO_VERSION
  merges:
    - enterprise $ODOO_VERSION
```
This is my configuration for `/odoo/custom/ssh` directory.
This is `config` file.

```
Host github.com
  HostName github.com
  IdentityFile ~/.ssh/id_rsa
  StrictHostKeyChecking no
```

I updated `id_rsa` and `id_rsa.pub`.
Then, I run `ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts` for the `known_hosts`.
After that, I run `sudo docker-compose -f test.yaml up -d` and I got the error like below image.

<img width="1435" alt="Screenshot 2024-10-15 at 00 50 08" src="https://github.com/user-attachments/assets/dfec1910-ba07-465a-bc93-af73cabfc85d">

@qrtl

**Comment by pedrobaeza** [comment]
I think the SSH should be put on the host machine that is doing the git aggregation, in the user that performs it through Docker. I don't use enteprise since long, but I remember to use a deploy token as something that can be embedded in the repos.yaml so no extra config is needed.

**Comment by AungKoKoLin1997** [comment]
@pedrobaeza  I tried to use github token for enterprise repo like this.

```
./odoo:
  defaults:
    # Shallow repositories ($DEPTH_DEFAULT=1) are faster & thinner
    # You may need a bigger depth when merging PRs (use $DEPTH_MERGE
    # for a sane value of 100 commits)
    depth: $DEPTH_DEFAULT
  remotes:
    ocb: https://github.com/OCA/OCB.git
    odoo: https://github.com/odoo/odoo.git
    openupgrade: https://github.com/OCA/OpenUpgrade.git
  target: ocb $ODOO_VERSION
  merges:
    - ocb $ODOO_VERSION
    # Example of a merge of the PR with the number <PR>
    # - oca refs/pull/<PR>/head
./private:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    origin: https://my_token@github.com/odoo/enterprise.git
  merges:
    - origin $ODOO_VERSION
```
But I still get the error.
```
64.86 From https://github.com/OCA/OCB
64.87  * branch              18.0       -> FETCH_HEAD
64.87  * [new branch]        18.0       -> ocb/18.0
64.88 (INFO) [08:39:33] git_aggregator.repo  odoo   Checking repo status
65.69 (INFO) [08:39:34] git_aggregator.repo  odoo   Reset branch to ocb 18.0
66.98 (INFO) [08:39:36] git_aggregator.repo  odoo   Execute shell after commands
66.98 (INFO) [08:39:36] git_aggregator.repo  odoo   End aggregation of /opt/odoo/custom/src/odoo
66.98 (INFO) [08:39:36] git_aggregator.repo  private   Start aggregation of /opt/odoo/custom/src/private
66.98 (INFO) [08:39:36] git_aggregator.repo  private   Switch to branch _git_aggregated
66.99 fatal: not a git repository (or any of the parent directories): .git
66.99 (ERROR) [08:39:36] git_aggregator.repo  private   /opt/odoo/custom/src/private> error calling ['git', 'checkout', '-B', '_git_aggregated']
67.00 Traceback (most recent call last):
67.00   File "/usr/local/lib/python3.10/site-packages/git_aggregator/main.py", line 230, in aggregate_repo
67.00     repo.aggregate()
67.00   File "/usr/local/lib/python3.10/site-packages/git_aggregator/repo.py", line 180, in aggregate
67.00     self._switch_to_branch(self.target['branch'])
67.00   File "/usr/local/lib/python3.10/site-packages/git_aggregator/repo.py", line 298, in _switch_to_branch
67.00     self.log_call(['git', 'checkout', '-B', branch_name], cwd=self.cwd)
67.00   File "/usr/local/lib/python3.10/site-packages/git_aggregator/repo.py", line 160, in log_call
67.00     ret = callwith(cmd, **kw)
67.00   File "/usr/local/lib/python3.10/subprocess.py", line 369, in check_call
67.00     raise CalledProcessError(retcode, cmd)
67.00 subprocess.CalledProcessError: Command '['git', 'checkout', '-B', '_git_aggregated']' returned non-zero exit status 128.
67.04 Traceback (most recent call last):
67.05   File "/usr/local/bin/autoaggregate", line 170, in <module>
67.05     aggregate(REPOS_YAML)
67.05   File "/usr/local/bin/autoaggregate", line 72, in aggregate
67.05     check_call(
67.05   File "/usr/local/lib/python3.10/subprocess.py", line 369, in check_call
67.05     raise CalledProcessError(retcode, cmd)
67.05 subprocess.CalledProcessError: Command '['gitaggregate', '--expand-env', '--config', '/opt/odoo/custom/src/repos.yaml', '--log-level', 'INFO', '--jobs', '1', 'aggregate']' returned non-zero exit status 1.
67.07 Traceback (most recent call last):
67.08   File "/opt/odoo/common/build", line 30, in <module>
67.08     subprocess.check_call(command)
67.08   File "/usr/local/lib/python3.10/subprocess.py", line 369, in check_call
67.08     raise CalledProcessError(retcode, cmd)
67.08 subprocess.CalledProcessError: Command '/opt/odoo/common/build.d/100-repos-aggregate' returned non-zero exit status 1.
```
I thought private folder is for the private repos and I tried to use this for enterprise addons. Am I missing something about the workflow?

**Comment by chienandalu** [comment]
`./private` is reserved for local addons (customer custom code). Have you tried just using `./enterprise` as your folder prefix?

**Comment by AungKoKoLin1997** [comment]
> Have you tried just using `./enterprise` as your folder prefix?

@chienandalu I did and there is no issue when I run `sudo docker-compose -f test.yaml up -d` but I don't see my enterprise addons inside directory. And what I thought is even it is exist in the directory, do we have to update `odoo.conf` to add enterprise addon?

```
lin@localhost:~/doodba$ sudo docker exec -it 887302afa32f /bin/bash
odoo@doodbatesting:/opt/odoo$ ls
auto  common  custom  requirements.txt
odoo@doodbatesting:/opt/odoo$ cd custom/
odoo@doodbatesting:/opt/odoo/custom$ ls
build.d  conf.d  dependencies  entrypoint.d  src  ssh
odoo@doodbatesting:/opt/odoo/custom$ cd src/
odoo@doodbatesting:/opt/odoo/custom/src$ ls
addons.yaml  odoo  private  repos.yaml
```

My repos.yaml.
```
# See https://github.com/Tecnativa/doodba#optodoocustomsrcreposyaml
./odoo:
  defaults:
    # Shallow repositories ($DEPTH_DEFAULT=1) are faster & thinner
    # You may need a bigger depth when merging PRs (use $DEPTH_MERGE
    # for a sane value of 100 commits)
    depth: $DEPTH_DEFAULT
  remotes:
    ocb: https://github.com/OCA/OCB.git
    odoo: https://github.com/odoo/odoo.git
    openupgrade: https://github.com/OCA/OpenUpgrade.git
  target: ocb $ODOO_VERSION
  merges:
    - ocb $ODOO_VERSION
    # Example of a merge of the PR with the number <PR>
    # - oca refs/pull/<PR>/head
./enterprise:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    origin: https://my_github_token@github.com/odoo/enterprise.git
  merges:
    - origin $ODOO_VERSION
```

**Comment by pedrobaeza** [comment]
Why are you trying with test.yaml instead of devel? 

About the syntax with deploy tokens, are you sure the URL is right?

One thing is sure, you need to put other repo name (like `./enterprise`). `./private` can't be used.

**Comment by AungKoKoLin1997** [comment]
> Why are you trying with test.yaml instead of devel?

I am using for staging purpose.

> One thing is sure, you need to put other repo name (like ./enterprise). ./private can't be used.

I can run successfully with enterprise by using `./enterprise` in `repos.yaml` and adding `enterprise: ['*']` in `addons.yaml`.

**Comment by pedrobaeza** [comment]
OK, then that was the problem.

**Comment by AungKoKoLin1997** [comment]
@chienandalu @pedrobaeza Thanks for the help!

**Comment by AungKoKoLin1997** [comment]
Can I ask another question in here?
I was not very clear about the intention of `/odoo/custom/src/private`. My understanding was this is for the customer custom addons and the addons under private folder will be linked automatically with odoo/addons as per this explanation from `addons.yaml`.

> Addons under private and odoo/addons are linked automatically unless you specify them.

Then, I found 

> Recommendation: use [repos.yaml](https://github.com/Tecnativa/doodba#optodoocustomsrcreposyaml) for everything except for [private](https://github.com/Tecnativa/doodba#optodoocustomsrcprivate), and ignore in your `.gitignore` and `.dockerignore` files every folder here except [private](https://github.com/Tecnativa/doodba#optodoocustomsrcprivate)

I am confusing how I should add my customer custom addons. Should I add like below using `addons.yaml`?
```
# Custom repositories
ENV:
  DEFAULT_REPO_PATTERN: https://github.com/qrtl/{}.git
  ODOO_VERSION: 16.0
some-repo:['*']
```
Or
Do I have to clone manually inside `private` folder?
Am I missing something about `private` folder?

**Comment by chienandalu** [comment]
You don't need to declare the private addons in the addons.yaml or repos.yaml. Just put their files in the private folder.

**Comment by AungKoKoLin1997** [comment]
@chienandalu 
I created my doodba project.
Then, put my customer addons like below.
```
/odoo/custom/src/private/my-private-repo
  ├── module_1/
      ├── __manifest__.py
      ├── __init__.py
      ├── models/
      ├── views/
  ├── module_2/
      ├── __manifest__.py
      ├── __init__.py
      ├── models/
      ├── views/
  ├── module_3/
      ├── __manifest__.py
      ├── __init__.py
      ├── models/
      ├── views/
```
Then, I run `sudo docker-compose -f test.yaml up -d` but I don't see my modules in the odoo instance's apps list.
I also checked inside docker `/opt/odoo/custom/src/private` and I can see my addons.

And I am curious about why does doodba put odoo code inside the docker.
If we put odoo code inside the docker,
- We need to get inside the Docker container to inspect the Odoo code
- Docker image gets larger than otherwise
- Odoo code update process gets a bit cumbersome as we need to update the Docker image
